src/glb-director-xdp/bpf/glb_encap.c

```diff
@@ -0,0 +1,367 @@
+	eth_hdr->ether_type = htons(PDNET_ETHER_TYPE_IPV4);
+
+	struct pdnet_ipv4_hdr *ipv4_hdr = (struct pdnet_ipv4_hdr *)(eth_hdr + 1);
+#ifdef EBPF_VERIFIER_GUARD
+	if ((void *)(ipv4_hdr + 1) > data_end) return XDP_DROP;
+#endif
+
+	config_bit = 1;
+	uint32_t *src_ip = (uint32_t *)bpf_map_lookup_elem(&config_bits, &config_bit);
+	if (src_ip == NULL)
+		return XDP_DROP;
+	
+	glb_bpf_printk("  src_ip: 0x%x\n", *src_ip);
+
+	ipv4_hdr->version = PDNET_IPV4_VERSION;
+	ipv4_hdr->ihl = PDNET_IPV4_HEADER_LEN;
+	ipv4_hdr->dscp = 0;
+	ipv4_hdr->ecn = 0;
+	ipv4_hdr->total_length =
+	    htons(sizeof(struct pdnet_ipv4_hdr) + sizeof(struct pdnet_udp_hdr) +
+		  sizeof(struct glb_gue_hdr) + 
+		  (sizeof(uint32_t) * remaining_hop_count) +
+		  inner_ip_total_length);
+	ipv4_hdr->identification = 0;
+	ipv4_hdr->fragment_offset = htons(PDNET_IPV4_FLAG_DF);
+	ipv4_hdr->time_to_live = PDNET_IPV4_DEFAULT_TTL;
+	ipv4_hdr->next_proto = PDNET_IP_PROTO_UDP;
+	ipv4_hdr->checksum = 0;
+	ipv4_hdr->src_addr = *src_ip;
+	ipv4_hdr->dst_addr = first_hop_ip;
+
+	ipv4_hdr->checksum = compute_ipv4_checksum(ipv4_hdr);
+
+	struct pdnet_udp_hdr *udp_hdr = (struct pdnet_udp_hdr *)(ipv4_hdr + 1);
+#ifdef EBPF_VERIFIER_GUARD
+	if ((void *)(udp_hdr + 1) > data_end) return XDP_DROP;
+#endif
+
+	/* Use the packet's entry hash and the low_hash (mostly TCP source port) to
+	 * generate the UDP source port. This ties each flow to an approximately 
+	 * random RX queue on the proxy hosts. Always set the high bit so we're 
+	 * using a very non-confusing (ephemeral) port number.
+	 */
+	udp_hdr->src_port = htons(0x8000 | ((pkt_hash ^ flow_hash) & 0x7fff));
+	udp_hdr->dst_port = htons(GLB_GUE_PORT);
+	udp_hdr->length =
+	    htons(sizeof(struct pdnet_udp_hdr) +
+	      sizeof(struct glb_gue_hdr) + 
+		  (sizeof(uint32_t) * remaining_hop_count) +
+		  inner_ip_total_length);
+	udp_hdr->checksum = 0;
+
+	struct glb_gue_hdr *gue_hdr = (struct glb_gue_hdr *)(udp_hdr + 1);
+
+#ifdef EBPF_VERIFIER_GUARD
+	if ((void *)(gue_hdr + 1) > data_end) return XDP_DROP;
+#endif
+
+	gue_hdr->private_type = 0;
+	gue_hdr->next_hop = 0;
+	gue_hdr->hop_count = remaining_hop_count;
+	/* hlen is essentially just private data, which is 1x 32 bits, plus the number of hops */
+	gue_hdr->version_control_hlen = 1 + remaining_hop_count;
+	gue_hdr->protocol = gue_ipproto;
+	gue_hdr->flags  = 0;
+#ifdef EBPF_VERIFIER_GUARD
+	if ((void *)&gue_hdr->hops[remaining_hop_count] > data_end) return XDP_DROP;
+#endif
+	/* hops are already encoded in network byte order (same as first_hop_ip above) */
+	/* unrolled manually to make less jumpy/cleaner eBPF output */
+	if (remaining_hop_count >= 1 && (void *)(&gue_hdr->hops[1]) <= data_end) gue_hdr->hops[0] = route_context->ipv4_hops[1];
+	if (remaining_hop_count >= 2 && (void *)(&gue_hdr->hops[2]) <= data_end) gue_hdr->hops[1] = route_context->ipv4_hops[2];
+	if (remaining_hop_count >= 3 && (void *)(&gue_hdr->hops[3]) <= data_end) gue_hdr->hops[2] = route_context->ipv4_hops[3];
```

Comment:
4 the max possible hop count in general or just in the xdp implementation? Don't remember seeing that limit before.

Reply:
The current implementation (and this one) has a maximum of 2 hops per row, and the ability to have an "old" and "current" hash component setting, making 4 total maximum hops.